core/src/main/java/io/questdb/griffin/engine/functions/analytic/RankFunctionFactory.java

```diff
@@ -0,0 +1,301 @@
+/*******************************************************************************
+ *     ___                  _   ____  ____
+ *    / _ \ _   _  ___  ___| |_|  _ \| __ )
+ *   | | | | | | |/ _ \/ __| __| | | |  _ \
+ *   | |_| | |_| |  __/\__ \ |_| |_| | |_) |
+ *    \__\_\\__,_|\___||___/\__|____/|____/
+ *
+ *  Copyright (c) 2014-2019 Appsicle
+ *  Copyright (c) 2019-2022 QuestDB
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ *
+ ******************************************************************************/
+
+package io.questdb.griffin.engine.functions.analytic;
+
+import io.questdb.cairo.*;
+import io.questdb.cairo.map.*;
+import io.questdb.cairo.sql.Record;
+import io.questdb.cairo.sql.*;
+import io.questdb.griffin.FunctionFactory;
+import io.questdb.griffin.SqlException;
+import io.questdb.griffin.SqlExecutionContext;
+import io.questdb.griffin.engine.RecordComparator;
+import io.questdb.griffin.engine.analytic.AnalyticContext;
+import io.questdb.griffin.engine.analytic.AnalyticFunction;
+import io.questdb.griffin.engine.functions.LongFunction;
+import io.questdb.griffin.engine.functions.constants.LongConstant;
+import io.questdb.griffin.engine.orderby.RecordComparatorCompiler;
+import io.questdb.std.*;
+
+public class RankFunctionFactory implements FunctionFactory {
+
+    @Override
+    public String getSignature() {
+        return "rank()";
+    }
+
+    @Override
+    public Function newInstance(int position, ObjList<Function> args, IntList argPositions, CairoConfiguration configuration, SqlExecutionContext sqlExecutionContext) {
+        final AnalyticContext analyticContext = sqlExecutionContext.getAnalyticContext();
+
+        if (analyticContext.getPartitionByRecord() != null) {
+            ArrayColumnTypes arrayColumnTypes = new ArrayColumnTypes();
+            arrayColumnTypes.add(ColumnType.LONG); // max index
+            arrayColumnTypes.add(ColumnType.LONG); // current index
+            arrayColumnTypes.add(ColumnType.LONG); // offset
+            Map map = MapFactory.createMap(configuration, analyticContext.getPartitionByKeyTypes(), arrayColumnTypes);
+            return new RankFunction(map, analyticContext.getPartitionByRecord(), analyticContext.getPartitionBySink());
+        } else if (analyticContext.isOrdered()) {
+            return new OrderRankFunction();
+        } else if (analyticContext.isOvered()) {
```

Comment:
Why do we need this new `isOvered()` method? We could assume that OVER clause is empty if there is no partition by record and order by. Am I missing anything?

Reply:
If use `rank()`, analyticContext properties like this:
<img width="379" alt="image" src="https://user-images.githubusercontent.com/5343547/196447938-a85dea97-59db-41f8-b2d2-4677f63c2dae.png">

If use `rank() over()`, analyticContext properties like this:
<img width="409" alt="image" src="https://user-images.githubusercontent.com/5343547/196447487-fa584130-981d-44c2-9382-11cbafc9b506.png"> 
the property `parititonByRecord` and `isOrder` are same.

Reply:
More informations:
<img width="606" alt="image" src="https://user-images.githubusercontent.com/5343547/196449930-eafbba5c-7dad-4b4f-9b59-f75225ad657c.png">

Reply:
But we support no over clause analytic function  in questdb already,  e.g.
```
select row_number() from xx;
```

Reply:
> But we supported no over clause analytic function in questdb already

I don't see any value in such behavior. To me, both `select rank() from x` and `select row_number() from x` should return an error, just like in Postgres.

@bluestreak01 @javier WDYT?

Reply:
I changed this logic in `RankFunctionFactory`, use `SequenceRowNumberFunction` to support `rank()` and `rank() over()`